packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,159 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import { createRule, getParserServices } from '../util';
+
+export default createRule({
+  defaultOptions: [],
+  meta: {
+    docs: {
+      description: 'Disallow type parameters that only appear once',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      sole: 'Type parameter {{name}} is used only once.',
+    },
+    schema: [],
+    type: 'problem',
+  },
+  name: 'no-unnecessary-type-parameters',
+  create(context) {
+    const parserServices = getParserServices(context);
+    let usage: Map<ts.Identifier, tsutils.VariableInfo> | undefined;
+
+    return {
+      [[
+        'ClassDeclaration[typeParameters]',
+        'ClassExpression[typeParameters]',
+        'ArrowFunctionExpression[typeParameters]',
+        'FunctionDeclaration[typeParameters]',
+        'FunctionExpression[typeParameters]',
+        'TSCallSignatureDeclaration[typeParameters]',
+        'TSConstructorType[typeParameters]',
+        'TSDeclareFunction[typeParameters]',
+        'TSFunctionType[typeParameters]',
+        'TSMethodSignature[typeParameters]',
+        'TSEmptyBodyFunctionExpression[typeParameters]',
+      ].join(', ')](esNode: TSESTree.FunctionLike): void {
+        const tsNode = parserServices.esTreeNodeToTSNodeMap.get(esNode) as
+          | ts.SignatureDeclaration
+          | ts.ClassLikeDeclaration;
+        if (!tsNode.typeParameters) {
+          return;
+        }
+
+        const checker = parserServices.program.getTypeChecker();
+
+        // XXX this collects a lot more usage than is needed for this rule.
+        usage ??= tsutils.collectVariableUsage(tsNode.getSourceFile());
+
+        // We need to resolve and analyze the inferred return type of a function
+        // to see whether it contains additional references to the type parameters.
+        // For classes, we need to do this for all their methods.
+        const inferredCounts = countInferredTypeParameterUsage(checker, tsNode);
+
+        // References to the type parameter in the function body are irrelevant
+        // for a valid type signature.
+        let declEndPos = tsNode.end;
+        if ('body' in tsNode) {
```

Comment:
[Request] We generally don't use `in` on AST nodes. The pattern is susceptible to missing new node types as they get added later. We generally prefer to do things like check `.type`, have different functions for different nodes, etc.

Or at least that's been my opinion. Filed #8226 to enforce this and/or determine whether I'm misleading folks.

Reply:
I've rewritten this as a nine clause conditional. TBH I think this is much less clear and much more problematic in terms of adding new node types. LMK what you think. In this case I really do care specifically whether the function-like thing has a body.